EOFTests/eof_validation/EOF1_eofcreate_invalid.json

```diff
@@ -0,0 +1,51 @@
+{
+  "EOF1_eofcreate_invalid": {
+    "vectors": {
+      "EOF1_eofcreate_invalid_0": {
+        "code": "0xef0001010004020001000903000100140400000000800004600060ff60006000ecef000101000402000100010400000000800000fe",
+        "results": {
+          "Prague": {
+            "exception": "EOF_TruncatedImmediate",
+            "result": false
+          }
+        }
+      },
+      "EOF1_eofcreate_invalid_1": {
+        "code": "0xef0001010004020001000a03000100140400000000800004600060ff60006000ec00ef000101000402000100010400000000800000fe",
+        "results": {
+          "Prague": {
+            "exception": "EOF_InvalidCodeTermination",
+            "result": false
+          }
+        }
+      },
+      "EOF1_eofcreate_invalid_2": {
+        "code": "0xef0001010004020001000c03000100140400000000800004600060ff60006000ec015000ef000101000402000100010400000000800000fe",
+        "results": {
+          "Prague": {
+            "exception": "EOF_InvalidContainerSectionIndex",
+            "result": false
+          }
+        }
+      },
+      "EOF1_eofcreate_invalid_3": {
+        "code": "0xef0001010004020001000c03000100140400000000800004600060ff60006000ecff5000ef000101000402000100010400000000800000fe",
+        "results": {
+          "Prague": {
+            "exception": "EOF_InvalidContainerSectionIndex",
+            "result": false
+          }
+        }
+      },
+      "EOF1_eofcreate_invalid_4": {
+        "code": "0xef0001010004020001000c03000100160400000000800004600060ff60006000ec005000ef000101000402000100010400030000800000feaabb",
+        "results": {
+          "Prague": {
+            "exception": "EOF_EofCreateWithTruncatedContainer",
```

Comment:
I do not see how we can detect this at validation time.  EOF Create with a truncated container should only fire if auxdata is insufficient, and that could change at runtime.  Although in this instance it is staticly true, we would have to add symbolic execution to the spec to prove it.

```
010004 # Types length ( 4 )
020001 # Total code sections ( 1 )
  000c # Code section 0 , 12 bytes
030001 # Total subcontainers ( 1 )
  0016 # Sub container 0, 22 byte
040000 # Data section length(  0 )
    00 # Terminator (end of header)
       # Code 0 types
    00 # 0 inputs 
    80 # 0 outputs  (Non-returning function)
  0004 # max stack:  4
       # Code section 0
  6000 # [0] PUSH1(0)
  60ff # [2] PUSH1(-1)
  6000 # [4] PUSH1(0)
  6000 # [6] PUSH1(0)
  ec00 # [8] EOFCREATE(0)
    50 # [10] POP
    00 # [11] STOP
           # Subcontainer 0 starts here
    ef0001 # Magic and Version ( 1 )
    010004 # Types length ( 4 )
    020001 # Total code sections ( 1 )
      0001 # Code section 0 , 1 bytes
    040003 # Data section length(  3 ) (actual size 2)    
        00 # Terminator (end of header)
           # Code 0 types
        00 # 0 inputs 
        80 # 0 outputs  (Non-returning function)
      0000 # max stack:  0
           # Code section 0
        fe # [0] INVALID
      # Data section length ( 3 ) actual length ( 2 )
      aabb
           # Subcontainer 0 ends
       # Data section (empty)
       ```

Reply:
Auxdata is appended to deploy container, it's not relevant here. 

Initcontainer (the one targeted by EOFCREATE) must have actual data size equal to declared size, otherwise this EOFCREATE is invalid.

In the spec:

> `EOFCREATE` the subcontainer pointed to by initcontainer_index must have its `len(data_section)` equal `data_size`, i.e. data section content is exactly as the size declared in the header (see [Data section lifecycle](https://github.com/ipsilon/eof/blob/main/spec/eof.md#data-section-lifecycle))

Reply:
First, I didn;t consider these may be the inputs in an InitContainer section, subject to AUXDATA from a TXCREATE opcode.  I was focused on contract creation transactions which have no such auxdata.

The spec appears to be the opposite of what was intended and needed then for EOFCREATE.  Consider a swap contract that includes the two contract addresses loaded by DATALOADN.  Those would need to point into the area that would be auxdata.  So for validation the subcontainer would need to have a data section length that would allow for the DATALOADN operation to validate (current length + 64 at least) but could not be set yet.  So the subcontainer can be "short" but after aux data is applied must not be "short" but can grow because of aux data.

Reply:
I think maybe you're missing one level of creation process nesting.

During creation this swap contract will be a subcontainer of initcontainer, this subcontainer is targeted by RETURNCONTRACT, truncated data is allowed for it, and is appended at run-time. Parent initcontainer is targeted by outer factory container EOFCREATE, truncated data is not allowed for it. 

Factory gets those constructor arguments from calldata probably, passes them to EOFCREATE, inside initcontainer execution they are forwarded to RETURNCONTRACT, and appended to deploy container data section.

Reply:
These two statements said different things...

> Initcontainer (the one targeted by EOFCREATE) must have actual data size equal to declared size, otherwise this EOFCREATE is invalid.

> During creation this swap contract will be a subcontainer of initcontainer, this subcontainer is targeted by RETURNCONTRACT, truncated data is allowed for it

The latter statement is what I was advocating for, so I will presume the first statement was made in error.  Even though it is different ops the InitContainer will be using AuxData to communicate the needed contract IDs as calldata may not be available as this may be a deeply nested create and not a top level create.  So initcode will need to be truncated similar to the deployable subcontainer.

Reply:
I think what Andrei wrote is correct, give me a moment I'll prepare an example to demonstrate this on the swap contract example

Reply:
@shemnon posted here, let me know if this clarifies things https://gist.github.com/pdobacz/27930c535fe891e51e77aa1baeee1572

Reply:
I think I see the disconnect.  Neither megaeof.md nor RIP-7620 specify what `data_offset` and `data_size` are used for.  The worked example drew that out:

* I had presumed they had the same semantics as `aux_data_offset` and `aux_data_size` and that the subcontainers are re-written with the data appended.
* Based on your example it appears that the data is copied from memory into calldata.

Neither one of these interpertations have any basis from either spec other than conventions of names.  Something like this needs to be added to https://eips.ethereum.org/EIPS/eip-7620#initcontainer-execution

> * Memory starting at `data_offset` of `data_size` length is copied into the call data

We also need verbiage in the opcode sections stating that 

> * Deduct gas for possible memory expansion for data starting at `data_offset` and going for `data_size` length

Reply:
Ah, yes, indeed. I think this was skipped in megaspec, because it tacitly extrapolates CREATE/CREATE2, and then copied like that into 7620. Thanks for pointing this out and sorry about the confusion :bow: 

Definitely worth emphasizing, that **neither** EOFCREATE **nor** TXCREATE rewrite the initcontainers they reference. They only do so on deployed containers referenced on RETURNCONTRACT they end with. So `data_offset` & `data_size` from EOFCREATE & TXCREATE are a very different beast from `aux_data_offset` & `aux_data_size` from RETURNCONTRACT.

To make this distinction clear, let's use `input_offset` and `input_size` for EOFCREATE / TXCREATE, which coincides with EXT*CALL naming from EIP-7069, and not with `aux_data` stuff.